onnxscript/ir/passes/_remove_unused.py

```diff
@@ -0,0 +1,82 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Utilities for removing unused nodes the IR graph."""
+
+from __future__ import annotations
+
+from collections import deque
+
+import onnxscript.ir as ir
+from onnxscript.ir import Attr, Graph, Node, Value, _enums
+
+
+class RemoveUnused:
+    def __init__(self, graph_like: Graph):
+        self._graph = graph_like
+
+    def purge(self) -> None:
+        """Remove unused nodes in this graph (and all subgraphs) that do not contribute to main graph outputs."""
+        # 1. Initialize:
+        #   Gather all nodes from the graph and its subgraphs.
+        #   Initialize sets to keep track of visited graphs, values, and nodes.
+        # 2. BFS traversal:
+        #   Create a queue initialized with all output values of the main graph.
+        #   While there are values in the queue:
+        #     - Dequeue a value and retrieve its producer node.
+        #     - Mark the producer node as visited, if it hasn't been visited.
+        #     - Enqueue all output values of the attribute subgraphs of the producer node,
+        #       if they haven't been visited.
+        #     - Enqueue all input values of the producer node, if they haven't been visited.
+        # 3. Remove:
+        #   Remove all nodes that have not been marked as visited during the BFS traversal.
+
+        # Initialize
+        all_nodes: list[Node] = list(ir.traversal.RecursiveGraphIterator(self._graph))
+        visited_graphs: set[Graph] = set()
+        visited_values: set[Value] = set()
+        visited_nodes: set[Node] = set()
+
+        # BFS Traversal
+        queue: deque[Value] = deque()
+
+        def add_graph_output_values_to_queue(graph: Graph | None) -> None:
+            """Helper function to add all output values of a graph to the queue."""
+            if not graph or graph in visited_graphs:
+                return
+            visited_graphs.add(graph)
+            for output in graph.outputs:
+                if not output:
+                    continue
+                queue.append(output)
+                visited_values.add(output)
+
+        add_graph_output_values_to_queue(self._graph)
+
+        while queue:
```

Comment:
It should be possible to avoid the queue by looping over all nodes in the backward order (assuming the ir preserve a consistent order on the nodes +@justinchuby).